libshaderc_util/src/compiler.cc

```diff
@@ -102,6 +103,46 @@ EShMessages GetMessageRules(shaderc_util::Compiler::TargetEnv env,
 
 namespace shaderc_util {
 
+unsigned int GlslangInitializer::initialize_count_ = 0;
+std::mutex* GlslangInitializer::glslang_mutex_ = nullptr;
+
+GlslangInitializer::GlslangInitializer() {
+  static std::mutex first_call_mutex;
+
+  // If this is the first call, glslang_mutex_ needs to be created, but in
+  // thread safe manner.
+  {
+    const std::lock_guard<std::mutex> first_call_lock(first_call_mutex);
+    if (glslang_mutex_ == nullptr) {
+      glslang_mutex_ = new std::mutex();
+    }
+  }
+
+  const std::lock_guard<std::mutex> glslang_lock(*glslang_mutex_);
+
+  if (initialize_count_ == 0) {
+    glslang::InitializeProcess();
+  }
+
+  initialize_count_++;
+}
+
+GlslangInitializer::~GlslangInitializer() {
+  const std::lock_guard<std::mutex> glslang_lock(*glslang_mutex_);
+
+  initialize_count_--;
+
+  if (initialize_count_ == 0) {
+    glslang::FinalizeProcess();
+    // There is no delete for glslang_mutex_ here, because we cannot guarantee
+    // there isn't a caller waiting for glslang_mutex_ in GlslangInitializer().
+    //
+    // This means that this class does leak one std::mutex worth of memory after
+    // the final instance is destroyed, but this allows us to defer allocating
+    // and constructing until we are sure we need to.
```

Comment:
Can the mutex be destroyed here?

Reply:
No, because there might a thread currently blocked in GlslangInitializer waiting for it. Deleting a mutex while holding it is undefined behaviour, so we would have to release it before deleting, which introduced a possible race.

At the core of the issue is that glslang retains some of its data in the global scope, thus InitializeProcess() is a static function, not tied to the lifetime of any specific object, so we cannot make any guarantees that it will not need to be invoked in the future. We cannot tie the mutex to the shaderc state lifetime, because we cannot guarantee that two different threads in the same program may create separate instances of shaderc and thus end up in here.